shell/platform/fuchsia/flutter/pointer_injector_delegate_unittest.cc

```diff
@@ -0,0 +1,685 @@
+    // This cannot be a local variable because the delayed task below can
+    // execute after this function returns.
+    auto canceled = std::make_shared<bool>(false);
+    bool timed_out = false;
+    async::PostDelayedTask(
+        loop->dispatcher(),
+        [loop, canceled, &timed_out] {
+          if (*canceled) {
+            return;
+          }
+          timed_out = true;
+          loop->Quit();
+        },
+        timeout);
+    loop->Run();
+    loop->ResetQuit();
+
+    if (!timed_out) {
+      *canceled = true;
+    }
+    return timed_out;
+  }
+
+  bool RunLoopWithTimeoutOrUntil(fit::function<bool()> condition,
+                                 zx::duration timeout,
+                                 zx::duration step) {
+    const zx::time timeout_deadline = zx::deadline_after(timeout);
+
+    while (zx::clock::get_monotonic() < timeout_deadline &&
+           loop_.GetState() == ASYNC_LOOP_RUNNABLE) {
+      if (condition()) {
+        loop_.ResetQuit();
+        return true;
+      }
+
+      if (step == zx::duration::infinite()) {
+        // Performs a single unit of work, possibly blocking until there is work
+        // to do or the timeout deadline arrives.
+        loop_.Run(timeout_deadline, true);
+      } else {
+        // Performs work until the step deadline arrives.
+        RunGivenLoopWithTimeout(&loop_, step);
+      }
+    }
+
+    loop_.ResetQuit();
+    return condition();
+  }
+
+  void RunLoopUntil(fit::function<bool()> condition,
+                    zx::duration step = zx::msec(10)) {
+    RunLoopWithTimeoutOrUntil(std::move(condition), zx::duration::infinite(),
+                              step);
+  }
+
+  void SetUp() override {
+    auto view_ref_pair = scenic::ViewRefPair::New();
+
+    host_view_ref_ = std::move(view_ref_pair.view_ref);
+
+    fup_RegistryHandle registry;
+    registry_ = std::make_unique<MockInjectorRegistry>(registry.NewRequest());
+
+    fuv_ViewRef host_view_ref_clone;
+    fidl::Clone(host_view_ref_, &host_view_ref_clone);
+
+    is_flatland_ = GetParam();
+    pointer_injector_delegate_ = std::make_unique<PointerInjectorDelegate>(
+        std::move(registry), std::move(host_view_ref_clone), is_flatland_);
+  }
+
+  void CreateView(uint64_t view_id,
+                  std::optional<fuv_ViewRef> view_ref = std::nullopt) {
+    if (!is_flatland_) {
+      pointer_injector_delegate_->OnCreateView(view_id);
+    } else {
+      fuv_ViewRef ref;
+      if (view_ref.has_value()) {
+        ref = std::move(*view_ref);
+      } else {
+        auto view_ref_pair = scenic::ViewRefPair::New();
+        ref = std::move(view_ref_pair.view_ref);
+      }
+      pointer_injector_delegate_->OnCreateView(view_id, std::move(ref));
+    }
+  }
+
+  std::unique_ptr<PointerInjectorDelegate> pointer_injector_delegate_;
+  std::unique_ptr<MockInjectorRegistry> registry_;
+  fuv_ViewRef host_view_ref_;
+  bool is_flatland_ = false;
+
+ private:
+  async::Loop loop_;
```

Comment:
Prefer to use async::TestLoop here, along with a subloop, via #include <lib/async-testing/test_loop.h> from the SDK

It gives you RunLoopUntil & friends for free, and FIDL works, all in a very deterministic way.  It even allows you to control the loop time, to test exotic edge cases!  See flatland_connection_unittest.cc for example uasge

Reply:
Filed a bug for the same. fxb/104285